File: ReSharper.FSharp/src/Services.FSharp/src/Debugger/FSharpBreakpointVariantsProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using JetBrains.Annotations;
using JetBrains.DocumentModel;
using JetBrains.ProjectModel;
using JetBrains.ReSharper.Feature.Services.Debugger;
using JetBrains.ReSharper.Psi;
using JetBrains.ReSharper.Psi.FSharp;
using JetBrains.ReSharper.Psi.FSharp.Tree;
using JetBrains.ReSharper.Psi.FSharp.Util;
using JetBrains.ReSharper.Psi.Tree;
using JetBrains.Rider.Model;
using JetBrains.Util;
using JetBrains.Util.dataStructures.TypedIntrinsics;
using Range = Microsoft.FSharp.Compiler.Range;

namespace JetBrains.ReSharper.Feature.Services.FSharp.Debugger
{
  [Language(typeof(FSharpLanguage))]
  public class FSharpBreakpointVariantsProvider : IBreakpointVariantsProvider
  {
    private const string MultilineBreakpointTextSuffix = " ...";

    public List<BreakpointVariantModelBase> GetBreakpointVariants(IProjectFile file, int line, ISolution solution)
    {
      var fsFile = file.GetPrimaryPsiFile() as IFSharpFile;
      var parseResults = fsFile?.ParseResults;
      var document = file.ToSourceFile()?.Document;
      if (parseResults == null || document == null)
        return null;

      var breakpointVariants = new JetHashSet<BreakpointVariantModelBase>();
      var documentLine = (Int32<DocLine>) line;
      var lineEndOffset = document.GetLineEndOffsetWithLineBreak(documentLine);
      var token = fsFile.FindTokenAt(new TreeOffset(document.GetLineStartOffset(documentLine)));
      while (token != null && token.GetTreeEndOffset().Offset < lineEndOffset)
      {
        var rangeOption = parseResults.ValidateBreakpointLocation(GetPos(document, token.GetTreeEndOffset().Offset));
        if (rangeOption == null || rangeOption.Value.StartLine - 1 != line)
        {
          token = token.GetNextToken();
          continue;
        }

        var range = rangeOption.Value;
        var startOffset = FSharpRangeUtil.GetStartOffset(document, range).Offset;
        var endOffset = FSharpRangeUtil.GetEndOffset(document, range).Offset;
        var breakpointLineText = document.GetText(new TextRange(startOffset, Math.Min(lineEndOffset, endOffset)));
        var breakpointText = endOffset > lineEndOffset
          ? breakpointLineText + MultilineBreakpointTextSuffix
          : breakpointLineText;

        breakpointVariants.Add(new BreakpointVariantModel(startOffset, endOffset, breakpointText));
        token = token.GetNextToken();
      }
      return breakpointVariants.ToList();
    }

    private static Range.pos GetPos([NotNull] IDocument document, int offset)
    {
      var coords = document.GetCoordsByOffset(offset);
      return Range.mkPos((int) coords.Line + 1, (int) coords.Column);
    }
  }
}

File: ReSharper.FSharp/src/FSharp.Psi.Features/src/Debugger/FSharpBreakpointVariantsProvider.fs
namespace JetBrains.ReSharper.Plugins.FSharp.Services.Debugger

open System
open System.Collections.Generic
open FSharp.Compiler.Text
open JetBrains.ReSharper.Feature.Services.Debugger
open JetBrains.ReSharper.Plugins.FSharp
open JetBrains.ReSharper.Plugins.FSharp.Psi
open JetBrains.ReSharper.Plugins.FSharp.Util
open JetBrains.ReSharper.Psi
open JetBrains.ReSharper.Psi.Tree
open JetBrains.Util

module FSharpBreakpointVariantsProvider =
    let supportedFileExtensions =
        [| FSharpProjectFileType.FsExtension
           FSharpProjectFileType.MlExtension
           FSharpScriptProjectFileType.FsxExtension
           FSharpScriptProjectFileType.FsScriptExtension |]

[<Language(typeof<FSharpLanguage>)>]
type FSharpBreakpointVariantsProvider() =

    let [<Literal>] multilineSuffix = " ..."

    interface IBreakpointVariantsProvider with
        member x.GetSupportedFileExtensions() =
            FSharpBreakpointVariantsProvider.supportedFileExtensions :> _

        member x.GetBreakpointVariants(file, line, _) =
            match file.GetPrimaryPsiFile().AsFSharpFile() with
            | null -> null
            | fsFile ->

            match fsFile.ParseResults with
            | None -> null
            | Some parseResults ->

            let document = file.ToSourceFile().Document
            let lineStart = document.GetLineStartOffset(docLine line)
            let lineEnd = document.GetLineEndOffsetWithLineBreak(docLine line)

            let result = Dictionary<range, IBreakpoint>() 
            for token in fsFile.FindTokensAt(TreeTextRange(TreeOffset(lineStart), TreeOffset(lineEnd))) do
                let documentEndOffset = token.GetDocumentEndOffset()
                let pos = getPosFromDocumentOffset documentEndOffset
                match parseResults.ValidateBreakpointLocation(pos) with
                | Some range when range.StartLine - 1 = line ->
                    if result.ContainsKey(range) then () else

                    let startOffset = getStartOffset document range
                    let endOffset = getEndOffset document range

                    let text =
                        let breakpointText = document.GetText(TextRange(startOffset, Math.Min(lineEnd, endOffset)))
                        if endOffset > lineEnd then breakpointText + multilineSuffix else breakpointText

                    // Multi-method breakpoints allow us to set up multiple breakpoints across multiple methods if they
                    // all point to the same place in the source code. This is the case e.g. for async CE (since there
                    // may be more than one function generated for a particular CE call).
                    result.[range] <-
                        TextRangeBreakpoint(TextRange(startOffset, endOffset), text, containingFunctionName = null,
                            isMultiMethodBreakpoint = true)
                | _ -> ()

            result.Values :> _
